### PR TITLE
bind9 9.17.9 patch

### PR DESCRIPTION
**Pull request by julek-wolfssl**

### CODE REVIEW

File: bind9/9.11.22.patch [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+ #endif
++#endif
+ 
+ #define DST_RET(a) {ret = a; goto err;}
+ 
+@@ -506,10 +508,10 @@ opensslecdsa_tofile(const dst_key_t *key, const char *directory) {
+ 	ret = dst__privstruct_writefile(key, &priv, directory);
+ 
+  err:
+-	if (eckey != NULL)
```
Comment (SparkiDev) [inline bind9/9.11.22.patch:510]: Why was this change needed?

File: bind9/9.11.22.patch [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+ 
+diff --git a/lib/isc/aes.c b/lib/isc/aes.c
+index 2ca07f6a4b..23060202a2 100644
+--- a/lib/isc/aes.c
++++ b/lib/isc/aes.c
+@@ -27,17 +27,21 @@
+ #include <openssl/opensslv.h>
+ #include <openssl/evp.h>
+ 
++#ifndef HAVE_WOLFSSL
```
Comment (SparkiDev) [inline bind9/9.11.22.patch:661]: Merge with #If below?
And other instances below.

File: bind9/9.11.22.patch [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+ 	int len;
+diff --git a/lib/isc/include/isc/util.h b/lib/isc/include/isc/util.h
+index 98101c981e..dcdff558e7 100644
+--- a/lib/isc/include/isc/util.h
++++ b/lib/isc/include/isc/util.h
+@@ -289,7 +289,7 @@ extern void mock_assert(const int result, const char* const expression,
+ /*% Unexpected Error */
+ #define UNEXPECTED_ERROR		isc_error_unexpected
+ /*% Fatal Error */
+-#define FATAL_ERROR			isc_error_fatal
```
Comment (SparkiDev) [inline bind9/9.11.22.patch:709]: Messy!
Could defining FATAL_ERROR to WS_FATAL_ERROR before including wolfssl headers and then undefing it after work any cleaner?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: bind9/9.11.22.patch (hunk 1 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:510]: This needed to be moved down because bind first free's the object and then reads from it. In OpenSSL this is fine because `EVP_PKEY_get1_EC_KEY` ups the ref counter. We create a new object and return it. This results in accessing freed memory.

File: bind9/9.11.22.patch (hunk 2 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:661]: How would I do this? These guards eliminate the stack allocations of `EVP_CIPHER_CTX`.

File: bind9/9.11.22.patch (hunk 3 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:709]: This would need to be done for every time that this gets included.
```
julek@julek-MS-7A38:~/Projects/bind9$ grep -rP '#include.*openssl/' --include=*.[ch] -l | wc -l
21
```
I don't think that this will be cleaner than changing all of the definitions in bind.

File: bind9/9.11.22.patch [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
++	then
++		echo "    wolfSSL cryptography (--with-wolfssl)"
++	else
++		test "X$CRYPTO" = "X" -o "yes" = "$want_native_pkcs11" || \
++			echo "    OpenSSL cryptography/DNSSEC (--with-openssl)"
++	fi
+ 	test "X$PYTHON" = "X" || echo "    Python tools (--with-python)"
+ 	test "X$XMLSTATS" = "X" || echo "    XML statistics (--with-libxml2)"
+ 	test "X$JSONSTATS" = "X" || echo "    JSON statistics (--with-libjson)"
+diff --git a/doc/dev/unexpected b/doc/dev/unexpected
```
Comment (haydenroche5) [inline bind9/9.11.22.patch:219]: It looks like this file doesn't exist in 9.11.22.

File: bind9/9.11.22.patch [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
++		PKG_CHECK_VAR(
++			[WOLFSSL_INCLUDEDIR],
++			[wolfssl],
++			[includedir],
++			[],
++			[AC_MSG_ERROR([Could not find wolfSSL includedir variable.])]
++		)
++		WOLFSSL_CFLAGS="${WOLFSSL_CFLAGS} -I${WOLFSSL_INCLUDEDIR}/wolfssl"
++	else	
++		WOLFSSL_CFLAGS="${WOLFSSL_CFLAGS} -I${use_wolfssl} -I${use_wolfssl}/wolfssl"
```
Comment (haydenroche5) [inline bind9/9.11.22.patch:139]: I think this should be:
`WOLFSSL_CFLAGS="${WOLFSSL_CFLAGS} -I${use_wolfssl}/include -I${use_wolfssl}/include/wolfssl"`
Since `$use_wolfssl` will be an installation prefix, like `/usr/local`.

Comment (haydenroche5) [reply bind9/9.11.22.patch:139]: I also needed to make this change to get it to work: `WOLFSSL_LIBS="${WOLFSSL_LIBS} -L${use_wolfssl}/lib -lwolfssl"`

**Comment by haydenroche5** [comment]
It seems like some relevant tests are being skipped when I run `make check`:

```
hayden@hayden-dell: /home/hayden/lenovo/bind-9.11.22$ grep -A3 "cryptography" errors.txt
I:This test requires support for cryptography
I:configure with --with-openssl, or --enable-native-pkcs11 --with-pkcs11
I:dnssec:Prerequisites missing, skipping test.
R:dnssec:SKIPPED
--
I:This test requires support for cryptography
I:configure with --with-openssl, or --enable-native-pkcs11 --with-pkcs11
I:rpzrecurse:Prerequisites missing, skipping test.
R:rpzrecurse:SKIPPED
--
I:This test requires support for cryptography
I:configure with --with-openssl, or --enable-native-pkcs11 --with-pkcs11
I:autosign:Prerequisites missing, skipping test.
R:autosign:SKIPPED
```

File: bind9/9.11.22.patch (hunk 5 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:139]: You are correct. Changing.

File: bind9/9.11.22.patch (hunk 4 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:219]: Are you using the gitlab source or a release bundle?

**Comment by julek-wolfssl** [comment]
I'm getting missing support for EDDSA cryptography (not supported in EVP layer) and GOST cryptography (not supported at all). I am getting a bunch of failures so I'll work on those.

File: bind9/9.11.22.patch (hunk 4 above)
Comment (haydenroche5) [reply bind9/9.11.22.patch:219]: Release bundle via https://downloads.isc.org/isc/bind9/9.11.22/. I can see how the customer is pulling down the source in their Yocto environment, but my guess is release bundle. Will post an update shortly.

**Comment by julek-wolfssl** [comment]
The same tests fail for both OpenSSL and wolfSSL so I'm not going to investigate any more.

File: bind9/9.11.22.patch (hunk 4 above)
Comment (haydenroche5) [reply bind9/9.11.22.patch:219]: The Yocto build is pulling the source code from https://ftp.isc.org/isc/bind9/9.11.22/bind-9.11.22.tar.gz, so ideally the patch should apply cleanly onto that, rather than the git repo.

Comment (julek-wolfssl) [reply bind9/9.11.22.patch:219]: Does the missing file make the patch fail? Or does it just issue a warning?

Comment (haydenroche5) [reply bind9/9.11.22.patch:219]: It fails. It causes an error and kills the build at the `do_patch` stage in the Yocto environment.

Comment (julek-wolfssl) [reply bind9/9.11.22.patch]: File removed.

File: bind9/9.11.22.patch [hunk 6]
```diff
@@ -0,0 +1,926 @@
...
+ }
+diff --git a/configure.ac b/configure.ac
+index 33bd76ac3b..b806df4360 100644
+--- a/configure.ac
++++ b/configure.ac
+@@ -1473,6 +1473,70 @@ AC_ARG_WITH(openssl,
+ 			   [Build with OpenSSL [yes|no|path].
+ 				(Crypto is required for DNSSEC)]),
+ 	    use_openssl="$withval", use_openssl="auto")
++
```
Comment (haydenroche5) [inline bind9/9.11.22.patch:109]: Just cosmetic, but I get whitespace errors when applying this with git apply:
```
hayden@hayden-dell: /home/hayden/tmp/bind-9.11.22$ git apply 9.11.22.patch
9.11.22.patch:109: trailing whitespace.
	    
9.11.22.patch:112: trailing whitespace.
#   
9.11.22.patch:113: trailing whitespace.
AC_ARG_WITH([wolfssl], 
9.11.22.patch:122: trailing whitespace.
	then	
9.11.22.patch:137: trailing whitespace.
	else	
warning: 5 lines add whitespace errors.
```

File: bind9/9.11.22.patch [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+ 	BN_GENCB *cb;
+ #if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
+ 	BN_GENCB _cb;
+@@ -460,7 +464,7 @@ openssldsa_generate(dst_key_t *key, int unused, void (*callback)(int)) {
+ 	if (result != ISC_R_SUCCESS)
+ 		return (result);
+ 
+-#if OPENSSL_VERSION_NUMBER > 0x00908000L
++#if OPENSSL_VERSION_NUMBER > 0x00908000L && !defined(HAVE_WOLFSSL)
+ 	dsa = DSA_new();
```
Comment (haydenroche5) [inline bind9/9.11.22.patch:443]: Why are we not doing `DSA_new()` for wolfSSL? Won't the subsequent call to `DSA_generate_parameters_ex` (AKA `wolfSSL_DSA_generate_parameters_ex`) fail because `dsa` is NULL?

File: bind9/9.11.22.patch [hunk 8]
```diff
@@ -0,0 +1,926 @@
...
+ 	for (i = 0; i < priv.nelements; i++) {
+diff --git a/lib/dns/opensslecdsa_link.c b/lib/dns/opensslecdsa_link.c
+index e9ea5eada6..56b9a93a82 100644
+--- a/lib/dns/opensslecdsa_link.c
++++ b/lib/dns/opensslecdsa_link.c
+@@ -39,12 +39,14 @@
+ #include <openssl/ecdsa.h>
+ #include <openssl/bn.h>
+ 
++#ifndef HAVE_WOLFSSL
```
Comment (haydenroche5) [inline bind9/9.11.22.patch:484]: Both `NID_X9_62_prime256v1` and `NID_secp384r1` are defined in wolfssl/openssl/ec.h. What's the reason for gating this block of code on `#ifndef HAVE_WOLFSSL`?

File: bind9/9.11.22.patch (hunk 3 above)
Comment (haydenroche5) [reply bind9/9.11.22.patch:698]: I agree with Juliusz that Sean's approach isn't as clean, but I also think it's the better approach overall. I don't see Bind accepting these changes upstream if we're messing with their defines. So maybe Sean's approach is preferable.

**Review by haydenroche5 — CHANGES_REQUESTED** [review]
Looks mostly good. The only functional issue I saw with the DSA allocation I commented on.

File: bind9/9.11.22.patch [hunk 9]
```diff
@@ -0,0 +1,926 @@
...
+ 
+@@ -180,9 +181,10 @@ installdirs:
+ 	$(SHELL) ${top_srcdir}/mkinstalldirs ${DESTDIR}${mandir}/man5
+ 	$(SHELL) ${top_srcdir}/mkinstalldirs ${DESTDIR}${mandir}/man8
+ 
++# use soft links as hard links aren't allowed on shared vm disks
+ install:: named@EXEEXT@ lwresd@EXEEXT@ installdirs
+ 	${LIBTOOL_MODE_INSTALL} ${INSTALL_PROGRAM} named@EXEEXT@ ${DESTDIR}${sbindir}
+-	(cd ${DESTDIR}${sbindir}; rm -f lwresd@EXEEXT@; @LN@ named@EXEEXT@ lwresd@EXEEXT@)
++	(cd ${DESTDIR}${sbindir}; rm -f lwresd@EXEEXT@; @LN@ named@EXEEXT@ lwresd@EXEEXT@) -s
```
Comment (haydenroche5) [inline bind9/9.11.22.patch:84]: I think the `-s` here should be inside the parentheses, perhaps? I'm getting this error when trying to install on the client's server:

```
| /bin/bash: -c: line 0: syntax error near unexpected token `-s'
```

**Comment by haydenroche5** [comment]
I'm adding `ECDSA_SIG_get0` and `ECDSA_SIG_set0` for libssh2 with [this PR](https://github.com/wolfSSL/wolfssl/pull/4302). This breaks the bind build because `HAVE_ECDSA_SIG_GET0` isn't defined for wolfSSL. All that needs to be modified (once that PR is merged) for this to work is an `AC_DEFINE` of `HAVE_ECDSA_SIG_GET0` in the wolfSSL section of configure.ac.

File: bind9/9.11.22.patch (hunk 3 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:698]: I'm honestly not seeing upstream accepting either approach. At least changing all instances is fairly self-explanatory for anyone looking to port other bind versions.

File: bind9/9.11.22.patch (hunk 7 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:443]: I opted for Bind to go the `DSA_generate_parameters` route. Its in the `#else` section of this very macro.

File: bind9/9.11.22.patch (hunk 8 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch:484]: They're both defined as enumerations. The preprocessor doesn't recognize this and errors out.

File: bind9/9.11.22.patch (hunk 9 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch]: Moved the `-s` into the brackets.

File: bind9/9.11.22.patch (hunk 6 above)
Comment (julek-wolfssl) [reply bind9/9.11.22.patch]: Removed trailing whitespace.

File: bind9/9.11.22.patch (hunk 3 above)
Comment (haydenroche5) [reply bind9/9.11.22.patch:698]: Fair enough. I'll call this one resolved.

File: bind9/9.11.22.patch (hunk 7 above)
Comment (haydenroche5) [reply bind9/9.11.22.patch:443]: Ah, ok. Cool. Bad reading on my part.

File: bind9/9.11.22.patch (hunk 8 above)
Comment (haydenroche5) [reply bind9/9.11.22.patch:484]: Got it.  Thanks for the explanation.

**Review by haydenroche5 — APPROVED** [review]
All my feedback has been addressed.

**Review by SparkiDev — APPROVED** [review]

**Comment by julek-wolfssl** [comment]
@haydenroche5 Fixed the `ECDSA_SIG_get0` issue.

**Comment by lealem47** [comment]
@julek-wolfssl Do the bind9 build options change after the 'HAVE_ECDSA_SIG_GET0' macro is added? Still getting these make errors
```
CC       libdns_la-opensslecdsa_link.lo
In file included from opensslecdsa_link.c:17:
/home/lealem/testing_fork/jenkins-scripts/stable/nightly/osp/osp/install-dir/include/wolfssl/openssl/ecdsa.h:69:32: error: static declaration of ‘wolfSSL_ECDSA_SIG_get0’ follows non-static declaration
   69 | #define ECDSA_SIG_get0         wolfSSL_ECDSA_SIG_get0
      |                                ^~~~~~~~~~~~~~~~~~~~~~
opensslecdsa_link.c:55:1: note: in expansion of macro ‘ECDSA_SIG_get0’
   55 | ECDSA_SIG_get0(const ECDSA_SIG *sig, const BIGNUM **pr, const BIGNUM **ps) {
      | ^~~~~~~~~~~~~~
/home/lealem/testing_fork/jenkins-scripts/stable/nightly/osp/osp/install-dir/include/wolfssl/openssl/ecdsa.h:49:18: note: previous declaration of ‘wolfSSL_ECDSA_SIG_get0’ was here
   49 | WOLFSSL_API void wolfSSL_ECDSA_SIG_get0(const WOLFSSL_ECDSA_SIG*,
      |                  ^~~~~~~~~~~~~~~~~~~~~~
/home/lealem/testing_fork/jenkins-scripts/stable/nightly/osp/osp/install-dir/include/wolfssl/openssl/ecdsa.h:70:32: error: static declaration of ‘wolfSSL_ECDSA_SIG_set0’ follows non-static declaration
   70 | #define ECDSA_SIG_set0         wolfSSL_ECDSA_SIG_set0
      |                                ^~~~~~~~~~~~~~~~~~~~~~
opensslecdsa_link.c:65:1: note: in expansion of macro ‘ECDSA_SIG_set0’
   65 | ECDSA_SIG_set0(ECDSA_SIG *sig, BIGNUM *r, BIGNUM *s) {
      | ^~~~~~~~~~~~~~
/home/lealem/testing_fork/jenkins-scripts/stable/nightly/osp/osp/install-dir/include/wolfssl/openssl/ecdsa.h:51:17: note: previous declaration of ‘wolfSSL_ECDSA_SIG_set0’ was here
   51 | WOLFSSL_API int wolfSSL_ECDSA_SIG_set0(WOLFSSL_ECDSA_SIG*, WOLFSSL_BIGNUM*,
      |                 ^~~~~~~~~~~~~~~~~~~~~~
make[5]: *** [Makefile:1801: libdns_la-opensslecdsa_link.lo] Error 1
make[5]: Leaving directory '/home/lealem/testing_fork/jenkins-scripts/stable/nightly/osp/osp/bind9-build/lib/dns'
make[4]: *** [Makefile:1918: all-recursive] Error 1
make[4]: Leaving directory '/home/lealem/testing_fork/jenkins-scripts/stable/nightly/osp/osp/bind9-build/lib/dns'
```

**Comment by julek-wolfssl** [comment]
Did you apply the latest patch and run `autoreconf` before configuring? Which bind version are you compiling?

**Comment by lealem47** [comment]
Applied the 9.17.9 patch and ran autoreconf -ivf. Both commands ran successfully. The version of bind is the 66fc6c5a9e32c95 commit on bind9

**Comment by julek-wolfssl** [comment]
Ok so I missed those defines in 9.17.9. I'll add it and open a small pr. Good catch @lealem47 .
